File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/OpaqueMutexedString.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class OpaqueMutexedString {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;

    List<Object> selfEdges = List.of();
    

    static class OpaqueMutexedStringCleaner implements Runnable {

        MemorySegment segment;
        OpaqueMutexedStringCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.OpaqueMutexedString_destroy(this.segment);
        }
    }

    OpaqueMutexedString() {}
    OpaqueMutexedString(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    public static OpaqueMutexedString fromUsize(long number) {
        
        var numberNative = number;
        var nativeVal = somelib_h.OpaqueMutexedString_from_usize(numberNative);
        List<Object> selfEdges = List.of();
        
        
        
        var returnVal = new OpaqueMutexedString(nativeVal, selfEdges);
        var cleaner = new OpaqueMutexedString.OpaqueMutexedStringCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static OpaqueMutexedString borrowOther(OpaqueMutexedString other) {
        
        var otherNative = other.internal;
        var nativeVal = somelib_h.OpaqueMutexedString_borrow_other(otherNative);
        List<Object> selfEdges = List.of(other);
        
        
        
        var returnVal = new OpaqueMutexedString(nativeVal, selfEdges);
        
        return returnVal;
    }
    
    
    public void change(long number) {
        
        
        var numberNative = number;
        somelib_h.OpaqueMutexedString_change(internal, numberNative);
        
    }
    
    public OpaqueMutexedString borrow() {
        
        
        var nativeVal = somelib_h.OpaqueMutexedString_borrow(internal);
        List<Object> selfEdges = List.of(this);
        
        
        
        var returnVal = new OpaqueMutexedString(nativeVal, selfEdges);
        
        return returnVal;
    }
    
    public OpaqueMutexedString borrowSelfOrOther(OpaqueMutexedString other) {
        
        
        var otherNative = other.internal;
        var nativeVal = somelib_h.OpaqueMutexedString_borrow_self_or_other(internal, otherNative);
        List<Object> selfEdges = List.of(this, other);
        
        
        
        var returnVal = new OpaqueMutexedString(nativeVal, selfEdges);
        
        return returnVal;
    }
    
    public long getLenAndAdd(long other) {
        
        
        var otherNative = other;
        var nativeVal = somelib_h.OpaqueMutexedString_get_len_and_add(internal, otherNative);
        return nativeVal;
    }
    
    public String dummyStr() {
        
        try (var arena = Arena.ofConfined()) {
            
            
            var nativeVal = somelib_h.OpaqueMutexedString_dummy_str(arena, internal);
            return SliceUtils.readUtf8(nativeVal);
        }
    }
    
    public Utf16Wrap wrapper() {
        
        
        var nativeVal = somelib_h.OpaqueMutexedString_wrapper(internal);
        List<Object> selfEdges = List.of();
        
        
        
        var returnVal = new Utf16Wrap(nativeVal, selfEdges);
        var cleaner = new Utf16Wrap.Utf16WrapCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/OpaqueMutexedString.kt
package dev.diplomattest.somelib;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface OpaqueMutexedStringLib: Library {
    fun OpaqueMutexedString_destroy(handle: Pointer)
    fun OpaqueMutexedString_from_usize(number: FFISizet): Pointer
    fun OpaqueMutexedString_change(handle: Pointer, number: FFISizet): Unit
    fun OpaqueMutexedString_borrow(handle: Pointer): Pointer
    fun OpaqueMutexedString_borrow_other(other: Pointer): Pointer
    fun OpaqueMutexedString_borrow_self_or_other(handle: Pointer, other: Pointer): Pointer
    fun OpaqueMutexedString_get_len_and_add(handle: Pointer, other: FFISizet): FFISizet
    fun OpaqueMutexedString_dummy_str(handle: Pointer): Slice
    fun OpaqueMutexedString_wrapper(handle: Pointer): Pointer
    fun OpaqueMutexedString_to_unsigned_from_unsigned(handle: Pointer, input: FFIUint16): FFIUint16
}

class OpaqueMutexedString internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
)  {

    internal class OpaqueMutexedStringCleaner(val handle: Pointer, val lib: OpaqueMutexedStringLib) : Runnable {
        override fun run() {
            lib.OpaqueMutexedString_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<OpaqueMutexedStringLib> = OpaqueMutexedStringLib::class.java
        internal val lib: OpaqueMutexedStringLib = Native.load("somelib", libClass)
        @JvmStatic
        
        fun fromUsize(number: ULong): OpaqueMutexedString {
            
            val returnVal = lib.OpaqueMutexedString_from_usize(FFISizet(number));
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = OpaqueMutexedString(handle, selfEdges)
            CLEANER.register(returnOpaque, OpaqueMutexedString.OpaqueMutexedStringCleaner(handle, OpaqueMutexedString.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun borrowOther(other: OpaqueMutexedString): OpaqueMutexedString {
            
            val returnVal = lib.OpaqueMutexedString_borrow_other(other.handle);
            val selfEdges: List<Any> = listOf(other)
            val handle = returnVal 
            val returnOpaque = OpaqueMutexedString(handle, selfEdges)
            return returnOpaque
        }
    }
    
    fun change(number: ULong): Unit {
        
        val returnVal = lib.OpaqueMutexedString_change(handle, FFISizet(number));
        
    }
    
    fun borrow(): OpaqueMutexedString {
        
        val returnVal = lib.OpaqueMutexedString_borrow(handle);
        val selfEdges: List<Any> = listOf(this)
        val handle = returnVal 
        val returnOpaque = OpaqueMutexedString(handle, selfEdges)
        return returnOpaque
    }
    
    fun borrowSelfOrOther(other: OpaqueMutexedString): OpaqueMutexedString {
        
        val returnVal = lib.OpaqueMutexedString_borrow_self_or_other(handle, other.handle);
        val selfEdges: List<Any> = listOf(this) + listOf(other)
        val handle = returnVal 
        val returnOpaque = OpaqueMutexedString(handle, selfEdges)
        return returnOpaque
    }
    
    fun getLenAndAdd(other: ULong): ULong {
        
        val returnVal = lib.OpaqueMutexedString_get_len_and_add(handle, FFISizet(other));
        return (returnVal.toULong())
    }
    
    fun dummyStr(): String {
        
        val returnVal = lib.OpaqueMutexedString_dummy_str(handle);
            return PrimitiveArrayTools.getUtf8(returnVal)
    }
    
    fun wrapper(): Utf16Wrap {
        
        val returnVal = lib.OpaqueMutexedString_wrapper(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Utf16Wrap(handle, selfEdges)
        CLEANER.register(returnOpaque, Utf16Wrap.Utf16WrapCleaner(handle, Utf16Wrap.lib));
        return returnOpaque
    }
    
    fun toUnsignedFromUnsigned(input: UShort): UShort {
        
        val returnVal = lib.OpaqueMutexedString_to_unsigned_from_unsigned(handle, FFIUint16(input));
        return (returnVal.toUShort())
    }

}